File: 0509-fibonacci-number/0509-fibonacci-number.js
/**
 * @param {number} n
 * @return {number}
 */
let dp;

var fib = function(n) {
    dp = new Array(100005).fill(-1);
    return fibo(n);
};

function fibo(n){
    if(n == 0 || n == 1 ) return n;
    if(dp[n] != -1) return dp[n];
    return dp[n] = fibo(n-1) + fibo(n-2);
}

File: 0509-fibonacci-number/0509-fibonacci-number.java
class Solution {
    public static int fibo_bottom_up(int n){
        int[] dp = new int[100005];
        dp[0] = 0;
        dp[1] = 1;

        for(int i=2; i<=n ; i++){
            dp[i] = dp[i-1] + dp[i-2];
        }
        return dp[n];
    }

    public int fib(int n) {
        return fibo_bottom_up(n);
    }
}